format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedInputFormatProvider.java

```diff
@@ -49,16 +50,53 @@
 @Name(DelimitedInputFormatProvider.NAME)
 @Description(DelimitedInputFormatProvider.DESC)
 public class DelimitedInputFormatProvider extends PathTrackingInputFormatProvider<DelimitedInputFormatProvider.Conf> {
-  public static final PluginClass PLUGIN_CLASS = getPluginClass();
   static final String NAME = "delimited";
   static final String DESC = "Plugin for reading files in delimited format.";
+  public static final PluginClass PLUGIN_CLASS = getPluginClass();
   private final Conf conf;
 
   public DelimitedInputFormatProvider(Conf conf) {
     super(conf);
     this.conf = conf;
   }
 
+  static Schema detectSchema(DelimitedConfig conf, String delimiter,
+                             InputFiles inputFiles, FormatContext context) throws IOException {
+    DataTypeDetectorStatusKeeper dataTypeDetectorStatusKeeper = new DataTypeDetectorStatusKeeper();
+    for (InputFile inputFile : inputFiles) {
+      String line;
+      String[] columnNames = null;
+      String[] rowValue;
+      try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputFile.open()))) {
+        for (int rowIndex = 0; rowIndex < conf.getSampleSize() && (line = reader.readLine()) != null; rowIndex++) {
+          rowValue = line.split(delimiter, -1);
+          if (rowIndex == 0) {
+            columnNames = DataTypeDetectorUtils.setColumnNames(line, conf.getSkipHeader(), conf.getEnableQuotedValues(),
+                                                               delimiter);
+            if (conf.getSkipHeader()) {
+              continue;
+            }
+          }
+          DataTypeDetectorUtils.detectDataTypeOfRowValues(conf.getOverride(), dataTypeDetectorStatusKeeper, columnNames,
+                                                          rowValue);
+        }
+        dataTypeDetectorStatusKeeper.validateDataTypeDetector();
+      }
+      List<Schema.Field> fields = DataTypeDetectorUtils.detectDataTypeOfEachDatasetColumn(
+        conf.getOverride(), columnNames, dataTypeDetectorStatusKeeper);
+      Schema schema = Schema.recordOf("text", fields);
+      return PathTrackingInputFormatProvider.addPathField(context.getFailureCollector(), schema, conf.getPathField());
+    }
+    return null;
+  }
+
+  private static PluginClass getPluginClass() {
```

Comment:
similar comment here, avoid moving methods around, it makes it hard to see what changed